File: smart-home-jd/src/hub/registry.ts
/**
 * The hub's persisted DEVICE REGISTRY - the single owner of every device's
 * static descriptor.
 *
 * Each record is published RETAINED on `smarthome/registry/<id>` so that:
 *   - the operator console / UIs always see the whole fleet, and
 *   - a sensor console can list and "take over" devices that are OFFLINE.
 * The records also persist to data/devices.json so the fleet survives restarts
 * (and a broker restart, which wipes retained messages).
 */
import { mkdirSync, readFileSync, writeFileSync, existsSync } from "node:fs";
import { join } from "node:path";
import type { MqttClient } from "mqtt";
import { publishJson, newId } from "../shared/mqtt.js";
import { topics } from "../shared/topics.js";
import {
  SENSOR_TYPE_LABELS,
  type DeviceRecord,
  type Location,
  type PairingAd,
  type SensorType,
} from "../shared/types.js";
import { DEVICE_CATALOG, makeSerial } from "../sensors/device-catalog.js";

function deriveName(type: SensorType, location?: Location): string {
  const label = SENSOR_TYPE_LABELS[type];
  return location ? `${label} - ${location.room}` : `${label} (unassigned)`;
}

export class Registry {
  private records = new Map<string, DeviceRecord>();
  private readonly file: string;

  constructor(
    private readonly client: MqttClient,
    private readonly dataDir: string,
  ) {
    this.file = join(dataDir, "devices.json");
  }

  /** Load persisted records, mark them offline (no live process yet), republish. */
  load(): void {
    let stored: DeviceRecord[] = [];
    try {
      if (existsSync(this.file)) {
        stored = JSON.parse(readFileSync(this.file, "utf8")) as DeviceRecord[];
      }
    } catch (err) {
      console.warn(`[hub] could not read ${this.file}: ${String(err)}`);
    }
    for (const rec of stored) {
      rec.status = "offline"; // a retained presence message will flip it back
      this.records.set(rec.deviceId, rec);
      this.publish(rec);
    }
    console.log(`[hub] registry loaded: ${this.records.size} device(s)`);
  }

  all(): DeviceRecord[] {
    return [...this.records.values()];
  }
  get(id: string): DeviceRecord | undefined {
    return this.records.get(id);
  }
  has(id: string): boolean {
    return this.records.has(id);
  }

  /**
   * Build and store a brand-new device. Used by both onboarding paths:
   *  - the direct register/ack shortcut (scripts, CLI `--type`), and
   *  - commissioning a device that was advertising in pairing mode.
   */
  private create(
    type: SensorType,
    opts: { deviceId?: string; serial?: string; location?: Location; name?: string } = {},
  ): DeviceRecord {
    const cat = DEVICE_CATALOG[type];
    let deviceId = opts.deviceId?.trim() || newId(type);
    if (this.records.has(deviceId)) deviceId = newId(type); // avoid collisions
    const location = opts.location;
    const record: DeviceRecord = {
      deviceId,
      type,
      name: opts.name?.trim() || deriveName(type, location),
      location,
      serialNumber: opts.serial ?? makeSerial(type),
      manufacturer: cat.manufacturer,
      model: cat.model,
      protocol: cat.protocol,
      powerSource: cat.powerSource,
      firmware: cat.firmware,
      capabilities: cat.capabilities,
      status: "offline",
      registeredAt: new Date().toISOString(),
    };
    this.records.set(deviceId, record);
    this.persist();
    this.publish(record);
    return record;
  }

  /** Direct registration (auto-commissioned) - scripts and CLI `--type`. */
  register(type: SensorType, desiredId?: string, location?: Location): DeviceRecord {
    const record = this.create(type, { deviceId: desiredId, location });
    console.log(`[hub] registered ${type} ${record.deviceId}`);
    return record;
  }

  /** Commission a device that advertised itself in pairing mode. */
  commission(ad: PairingAd, name?: string, location?: Location): DeviceRecord {
    const record = this.create(ad.type, { serial: ad.serial, name, location });
    console.log(`[hub] commissioned ${ad.type} ${record.deviceId} (serial ${ad.serial})`);
    return record;
  }

  /** Update liveness. Returns true if the status actually changed. */
  setStatus(id: string, status: "online" | "offline"): boolean {
    const rec = this.records.get(id);
    if (!rec) return false;
    const changed = rec.status !== status;
    rec.status = status;
    rec.lastSeen = new Date().toISOString();
    if (changed) {
      this.persist();
      this.publish(rec);
    }
    return changed;
  }

  /** Assign (or, with null, clear) a device's room. */
  assign(id: string, location: Location | null): void {
    const rec = this.records.get(id);
    if (!rec) return;
    rec.location = location ?? undefined;
    rec.name = deriveName(rec.type, rec.location);
    this.persist();
    this.publish(rec);
    console.log(
      `[hub] ${id} assigned -> ${rec.location ? rec.location.room : "(unassigned)"}`,
    );
  }

  remove(id: string): void {
    if (!this.records.delete(id)) return;
    this.persist();
    // Clear retained registry + presence so the device vanishes everywhere.
    this.client.publish(topics.registry(id), "", { retain: true, qos: 1 });
    this.client.publish(topics.presence(id), "", { retain: true, qos: 1 });
    console.log(`[hub] removed ${id}`);
  }

  private publish(rec: DeviceRecord): void {
    publishJson(this.client, topics.registry(rec.deviceId), rec, { retain: true });
  }

  private persist(): void {
    try {
      mkdirSync(this.dataDir, { recursive: true });
      writeFileSync(this.file, JSON.stringify(this.all(), null, 2));
    } catch (err) {
      console.warn(`[hub] could not write ${this.file}: ${String(err)}`);
    }
  }
}

File: smart-home-jd-lean/hub/crates/hub/src/registry.rs
//! The hub's persisted DEVICE REGISTRY: the single owner of every device's
//! static descriptor. Each record is published RETAINED on
//! `smarthome/registry/<id>` and persisted to `<dataDir>/devices.json`.
use std::collections::HashMap;
use std::path::PathBuf;

use rumqttc::AsyncClient;
use shared::util::{new_id, now_iso};
use shared::{topics, DeviceRecord, Location, SensorType, Status};

use crate::catalog::{catalog, make_serial};
use crate::mqtt::{clear_retained, publish_json};

fn derive_name(t: SensorType, location: Option<&Location>) -> String {
    match location {
        Some(l) => format!("{} - {}", t.label(), l.room),
        None => format!("{} (unassigned)", t.label()),
    }
}

pub struct Registry {
    records: HashMap<String, DeviceRecord>,
    client: AsyncClient,
    file: PathBuf,
    data_dir: PathBuf,
}

#[derive(Default)]
pub struct CreateOpts {
    pub device_id: Option<String>,
    pub serial: Option<String>,
    pub location: Option<Location>,
    pub name: Option<String>,
}

impl Registry {
    pub fn new(client: AsyncClient, data_dir: PathBuf) -> Self {
        let file = data_dir.join("devices.json");
        Registry {
            records: HashMap::new(),
            client,
            file,
            data_dir,
        }
    }

    /// Load persisted records, mark them offline, republish (retained).
    pub async fn load(&mut self) {
        let mut stored: Vec<DeviceRecord> = Vec::new();
        if self.file.exists() {
            match std::fs::read_to_string(&self.file) {
                Ok(s) => match serde_json::from_str::<Vec<DeviceRecord>>(&s) {
                    Ok(v) => stored = v,
                    Err(e) => eprintln!("[hub] could not parse {:?}: {e}", self.file),
                },
                Err(e) => eprintln!("[hub] could not read {:?}: {e}", self.file),
            }
        }
        for mut rec in stored {
            rec.status = Status::Offline; // a retained presence message flips it back
            let id = rec.device_id.clone();
            self.publish(&rec).await;
            self.records.insert(id, rec);
        }
        println!("[hub] registry loaded: {} device(s)", self.records.len());
    }

    pub fn all(&self) -> impl Iterator<Item = &DeviceRecord> {
        self.records.values()
    }
    pub fn get(&self, id: &str) -> Option<&DeviceRecord> {
        self.records.get(id)
    }
    pub fn len(&self) -> usize {
        self.records.len()
    }

    async fn create(&mut self, t: SensorType, opts: CreateOpts) -> DeviceRecord {
        let cat = catalog(t);
        let mut device_id = opts
            .device_id
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .unwrap_or_else(|| new_id(t.as_str()));
        if self.records.contains_key(&device_id) {
            device_id = new_id(t.as_str());
        }
        let name = opts
            .name
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .unwrap_or_else(|| derive_name(t, opts.location.as_ref()));
        let record = DeviceRecord {
            device_id: device_id.clone(),
            kind: t,
            name,
            location: opts.location,
            serial_number: opts.serial.unwrap_or_else(|| make_serial(t)),
            manufacturer: cat.manufacturer.to_string(),
            model: cat.model.to_string(),
            protocol: cat.protocol,
            power_source: cat.power_source,
            firmware: cat.firmware.to_string(),
            capabilities: cat.capabilities.iter().map(|s| s.to_string()).collect(),
            status: Status::Offline,
            registered_at: now_iso(),
            last_seen: None,
        };
        self.records.insert(device_id, record.clone());
        self.persist();
        self.publish(&record).await;
        record
    }

    /// Direct registration (auto-commissioned) - used by `ctl` / scripts.
    pub async fn register(
        &mut self,
        t: SensorType,
        desired_id: Option<String>,
        location: Option<Location>,
    ) -> DeviceRecord {
        let rec = self
            .create(
                t,
                CreateOpts {
                    device_id: desired_id,
                    location,
                    ..Default::default()
                },
            )
            .await;
        println!("[hub] registered {} {}", t.as_str(), rec.device_id);
        rec
    }

    /// Commission a device that advertised itself in pairing mode.
    pub async fn commission(
        &mut self,
        serial: String,
        kind: SensorType,
        name: Option<String>,
        location: Option<Location>,
    ) -> DeviceRecord {
        let rec = self
            .create(
                kind,
                CreateOpts {
                    serial: Some(serial.clone()),
                    name,
                    location,
                    ..Default::default()
                },
            )
            .await;
        println!(
            "[hub] commissioned {} {} (serial {})",
            kind.as_str(),
            rec.device_id,
            serial
        );
        rec
    }

    /// Update liveness. Returns true if the status actually changed.
    pub async fn set_status(&mut self, id: &str, status: Status) -> bool {
        let Some(rec) = self.records.get_mut(id) else {
            return false;
        };
        let changed = rec.status != status;
        rec.status = status;
        rec.last_seen = Some(now_iso());
        if changed {
            let rec = rec.clone();
            self.persist();
            self.publish(&rec).await;
        }
        changed
    }

    /// Assign (or, with None, clear) a device's room.
    pub async fn assign(&mut self, id: &str, location: Option<Location>) {
        let Some(rec) = self.records.get_mut(id) else {
            return;
        };
        rec.location = location;
        rec.name = derive_name(rec.kind, rec.location.as_ref());
        let rec = rec.clone();
        self.persist();
        self.publish(&rec).await;
        println!(
            "[hub] {id} assigned -> {}",
            rec.location.as_ref().map(|l| l.room.as_str()).unwrap_or("(unassigned)")
        );
    }

    pub async fn remove(&mut self, id: &str) {
        if self.records.remove(id).is_none() {
            return;
        }
        self.persist();
        clear_retained(&self.client, &topics::registry(id)).await;
        clear_retained(&self.client, &topics::presence(id)).await;
        println!("[hub] removed {id}");
    }

    async fn publish(&self, rec: &DeviceRecord) {
        publish_json(&self.client, &topics::registry(&rec.device_id), rec, true).await;
    }

    fn persist(&self) {
        if let Err(e) = std::fs::create_dir_all(&self.data_dir) {
            eprintln!("[hub] could not create {:?}: {e}", self.data_dir);
            return;
        }
        let all: Vec<&DeviceRecord> = self.records.values().collect();
        match serde_json::to_string_pretty(&all) {
            Ok(s) => {
                if let Err(e) = std::fs::write(&self.file, s) {
                    eprintln!("[hub] could not write {:?}: {e}", self.file);
                }
            }
            Err(e) => eprintln!("[hub] serialize registry failed: {e}"),
        }
    }
}
